File: app/src/main/java/com/jamieadkins/gwent/card/detail/DetailPresenter.java
package com.jamieadkins.gwent.card.detail;

import android.support.annotation.NonNull;

import com.jamieadkins.gwent.card.CardFilter;
import com.jamieadkins.gwent.data.CardDetails;
import com.jamieadkins.gwent.data.interactor.CardsInteractor;
import com.jamieadkins.gwent.data.interactor.RxDatabaseEvent;

import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Listens to user actions from the UI, retrieves the data and updates the
 * UI as required.
 */

public class DetailPresenter implements DetailContract.Presenter {
    private final CardsInteractor mDetailInteractor;
    private final DetailContract.View mDetailView;
    private String mCardId;

    public DetailPresenter(@NonNull DetailContract.View detailView,
                           @NonNull CardsInteractor detailInteractor) {
        mDetailInteractor = detailInteractor;

        mDetailView = detailView;
        mDetailView.setPresenter(this);
    }

    @Override
    public Single<RxDatabaseEvent<CardDetails>> getCard(String cardId) {
        return mDetailInteractor.getCard(cardId);
    }

    @Override
    public void start() {

    }

    @Override
    public void setCardId(String cardId) {
        mCardId = cardId;
    }

    @Override
    public String getCardId() {
        return mCardId;
    }

    @Override
    public void stop() {
        mDetailInteractor.removeListeners();
    }

    @Override
    public Observable<RxDatabaseEvent<CardDetails>> getCards(CardFilter cardFilter) {
        return mDetailInteractor.getCards(cardFilter);
    }

    @Override
    public void reportMistake(String cardId, String description) {
        mDetailInteractor.reportMistake(cardId, description);
    }
}

File: app/src/main/java/com/jamieadkins/gwent/card/detail/DetailPresenter.kt
package com.jamieadkins.gwent.card.detail

import com.jamieadkins.gwent.base.BaseDisposableObserver
import com.jamieadkins.gwent.domain.card.GetCardUseCase
import com.jamieadkins.gwent.domain.card.GetCardsUseCase
import com.jamieadkins.gwent.main.BasePresenter
import io.reactivex.subjects.BehaviorSubject
import javax.inject.Inject

class DetailPresenter @Inject constructor(
    private val view: DetailContract.View,
    private val getCardUseCase: GetCardUseCase,
    private val getCardsUseCase: GetCardsUseCase
) : BasePresenter(), DetailContract.Presenter {

    private val cardId = BehaviorSubject.create<String>()

    override fun onAttach() {
        cardId
            .switchMap { getCardUseCase.getCard(it) }
            .switchMap { card ->
                getCardsUseCase.getCards(card.relatedCards)
                    .map { related ->
                        CardDetailsScreenData(card, related.toList())
                    }
            }
            .doOnSubscribe { view.showLoadingIndicator() }
            .subscribeWith(object : BaseDisposableObserver<CardDetailsScreenData>() {
                override fun onNext(result: CardDetailsScreenData) {
                    view.showScreen(result)
                    view.hideLoadingIndicator()
                }
            })
            .addToComposite()
    }

    override fun setCardId(cardId: String) = this.cardId.onNext(cardId)
}
